File: leetcode/2024/cpp/131/partition.cpp
class Solution {
public:
    bool checkPal(string &s, int i, int j) {
        while (i < j) {
            if (s[i] != s[j])
                return false;
            i += 1;
            j -= 1;
        }
        return true;
    }
    void helper(int idx, vector<string> &temp, vector<vector<string>> &result, string &s) {
        if (idx == s.size()) {
            result.push_back(temp);
            return;
        }
        for(int i = idx; i < s.size(); i++){
            if(checkPal(s, idx, i)){
                temp.push_back(s.substr(idx, i - idx + 1));
                helper(i + 1, temp, result, s);
                temp.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        vector<string> temp;
        helper(0, temp, res, s);
        return res;
    }
};

File: leetcode/2024/python/131/partition.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = []
        part = []
        def isPal(s, i, j):
            while i < j:
                if s[i] != s[j]:
                    return False
                i += 1
                j -= 1
            return True
        def dfs(idx):
            if idx == len(s):
                res.append(part.copy())
                return
            for i in range(idx, len(s)):
                if isPal(s, idx, i):
                    part.append(s[idx:i+1])
                    dfs(i+1)
                    part.pop()
        dfs(0)
        return res
    
